[RDY] Use SDL_gpu library for faster SDL rendering

mutability opened the pull request:
Fixes #19071

This changes all the SDL code from using the SDL_render interface to using SDL_gpu (https://github.com/grimfang4/sdl-gpu), which is a separate library that provides a rendering abstraction on top of SDL. This obviously adds a new build dependency for the library, and it also currently requires support for OpenGL 2+ or better. That should be very common these days, anything with any sort of 3D acceleration will have OpenGL 2+ support.

SDL_gpu can be much faster if you are careful about how you arrange your blitting. In particular if you do many blits from the same texture to the same target, it will batch those together, which produces a huge speedup.

This PR first does the obvious changes to use SDL_gpu, and then implements a class that maintains an atlas texture that gathers up all the tiles in a tileset, or all the glyphs in a font, onto a single texture to take advantage of the blitting speedup.

Benchmarked here on a discrete Nvidia GTX 670M, and on the integrated graphics (Intel HD Graphics 4000). The SDL_gpu code is faster in all cases except one. The exception is when using tiles, maximum zoomed in, minimap on; the minimap code is currently somewhat slower in the SDL_gpu version and that cost dominates that case. The difference there is pretty small, 109fps vs 102fps.

In all other case the SDL_gpu code ranges from moderately faster with tiles (48fps vs 29fps for maximum-zoom-out, no minimap) to _dramatically_ faster for "SDL ASCII" (52fps vs 5fps, a 10x speedup)

Tested on Linux (Ubuntu 16.04) only. I have done a compile test for WIndows with MXE but I can't test the result as the Windows VM I have available doesn't have suitable opengl acceleration. @BrettDong tested OS X in #19071.

This needs testing on Windows and on a wider variety of video cards.

jasonmbrown commented:
Mutability can you link the compiled executable for windows builds? Im willing to test and let ya know the benchmark differences I get compared to normal build.

macrosblackd commented:
Looks interesting, but is there a way to build with Windows (not mxe)?

mutability commented:
Presumably it can be built for Windows via the other mechanisms, but I don't have those set up so I don't know the details. You need to (a) build SDL_gpu (which is cmake-based and I believe is set up to build with VS) and (b) add the resulting library to the set of libraries that C:DDA links.

mutability commented:
mxe build is here: https://www.dropbox.com/s/v6l6tcmdd8pnwtv/cdda-sdlgpu-win.zip?dl=0

I checked the curses build works OK, but the tiles version will not start in my VM (just hangs with a black window at the point where it tries to initialize opengl); I don't know whether than is a VM problem or a build problem.

If you get a window that pops up then immediately disappears, check the debug log in config/, any SDL_gpu init errors are logged there.

mutability commented:
Resurrected my real-hardware Windows install and the tiles version runs OK there, so it was just a VM problem.

jasonmbrown commented:
Comparisons Were Run on Clean Worlds and Saves tested using the debug Draw Benchmark command
They were done at Default Zoom, with all default settings.

SDL_GPU build:
SDL_GPU FPS 22 With Minimap
SDL_GPU FPS 55 Without Minimap
---
Comparison Build:
Version:0.C-19675-gb950780 (tiles)
Build:5896
Hardware Render:
41 FPS With Minimap
54 FPS Without Minimap

Software Render: 
8 FPS With Minimap
12.5 FPS Without Minimap

results:
GPU_SDL ran 50% slower for me With minimap.
GPU_SDL Ran Mildly faster Without minimap.

Notes:
Even though the boost Doesnt appear to be a Huge amount (probably due to my integrated GPU IntelHD 3000), This fixed all my Draw Related Menu lag (I had a good 2-3 second delay when using hardware render, on basically any menu that was displaying at the same time tiles were displaying).
Hopefully this gets added to the Experimental build soon.

mutability commented:
Was that with a tileset, or with "use tiles"=False?

mutability commented:
Here's my benchmark results:

tileset | zoom | minimap | nvidia sdlrender | nvidia sdlgpu | intel sdlrender | intel sdlgpu
-- | -- | -- | -- | -- | -- | --
chesthole | normal | yes | 76 | 83 | 43 | 59
chesthole | max | yes | 109 | 102 | 94 | 68
chesthole | min | yes | 27 | 41 | 8.1 | 35
none | n/a | yes | 5.4 | 52 | 4.6 | 42
chesthole | normal | no | 100 | 123 | 52 | 117
chesthole | max | no | 167 | 170 | 142 | 153
chesthole | min | no | 29 | 48 | 8.3 | 48
none | n/a | no | 5.5 | 65 | 4.7 | 65

BrettDong commented:
There are some problems displaying wide characters:

1. Blue selection cursor only appears on the left half side on each character in screen capture 1
2. The center prompt box in screen capture 1 seems transparent, the background ASCII art is visible
3. The wide characters are not redrawed completely and making words overlapped together in screen capture 2
4. A `-` mark appears after the period punctuation in screen capture 3 and I have no idea of the reason.

![2016-11-23 10 40 06](https://cloud.githubusercontent.com/assets/21075502/20565819/f1688d56-b1cd-11e6-9a7e-b57dd60b996d.png)
![2016-11-23 10 40 41](https://cloud.githubusercontent.com/assets/21075502/20565817/f0c13f10-b1cd-11e6-8c66-0647fafce426.png)
![2016-11-23 10 40 47](https://cloud.githubusercontent.com/assets/21075502/20565818/f0e270ae-b1cd-11e6-8b1b-92678ce136f4.png)

BrettDong commented:
![2016-11-23 10 47 26](https://cloud.githubusercontent.com/assets/21075502/20566106/e67176d2-b1ce-11e6-89ff-98a895f73779.png)

Another screenshot FYI

mutability commented:
I think all of those are manifestations of the same problem: only half of the background of a double-width glyph is drawn.

mutability commented:
Try that, I don't have a suitable font to test with here so it was all placeholders but it did seem to handle the double-width background correctly.

mutability commented:
Updated windows builds with the wide-character fix and a faster minimap:

https://www.dropbox.com/s/ici4v4ih3dan0k9/cdda-sdlgpu-win-5dc16ff6.zip?dl=0 (full package)
https://www.dropbox.com/s/p4a0ut4e6f64gq7/cataclysm-tiles-5dc16ff6.exe?dl=0 (just the exe)

mutability commented:
| tileset | zoom | minimap | intel sdlrender | intel sdlgpu e05c996 | intel sdlgpu 5dc16ff |
| -- | -- | -- | -- | -- | -- |
| chesthole | normal | yes | 43 | 59 | **86** |
| chesthole | max | yes | 94 | 68 | **102** |
| chesthole | min | yes | 8.1 | 35 | **42** |
| none | n/a | yes | 4.6 | 42 | **54** |

jasonmbrown commented:
Muta my benchmarks were done using chesthole tileset, and your linked bin from above.

BrettDong commented:
Nice, now work perfectly.


MacBook Air, macOS Sierra 10.12.1, Intel HD Graphics 6000

With minimap on: 74 FPS

Without minimap: 93.5 FPS

mutability commented:
@jasonmbrown can you check the updated build in my recent comment above and see if that helps performance on your system with the minimap on?

DeadLeaves commented:
This fixes a drawing bug I've been having on my laptop with Intel integrated graphics card. (Weird black lines between sprites) If also delivers on performance which I haven't had time to test, this seems great.

Zireael07 commented:
> This fixes a drawing bug I've been having on my laptop with Intel integrated graphics card. (Weird black lines between sprites)

That's great! The bug has been plaguing quite a number of players and no one could figure out what it was...

jasonmbrown commented:
@mutability I tested again with the more recent build you linked. ran 5 benchmarks got between 30fps and 34fps with the minimap and between 50fps and 60fps with it turned off. But All around it looks and feels alot smoother.  Im really looking forwards to this getting merged

patternoia commented:
Great work @mutability !
I need to warn you that if merged, this will break the macOS builds, because they are built with FRAMEWORK=1. 
`sdl-gpu` already supports building the `SDL2_gpu.framework` if build with `cmake -G "Unix Makefiles"`.
You would need to update the `Makefile` accordingly to link against this framework like it is now done for regular `SDL2.framework`

mutability commented:
I can't build on macOS (or Visual Studio, or native mingw/msys) so patches to the Makefile and/or COMPILING.md would be appreciated

patternoia commented:
@BrettDong since you are already into this PR, could you provide us such a patch?

mutability commented:
Except for build stuff I think this is ready to go.

A word of caution: the software rendering option no longer does anything. The tiles version will only work if you have OpenGL 2 or higher available. If we want to retain support for software rendering in some form, I have some work in progress to restore software rendering by adding a SDL_gpu renderer that maps back to SDL_render's software renderer, but that's not ready yet and I'd rather handle that that as a separate PR.

patternoia commented:
Do you mean an abstract renderer factory class which allows to switch between (and instantiate a proper instance of a) software renderer, accelerated SDL2 renderer or new SDL2_gpu? 
That would be a good non-breaking transition approach which will allow to obsolete others with time.

mutability commented:
No, hooking in another abstraction layer like that is pretty annoying. I mean everything uses SDL_gpu, and providing a SDL_gpu renderer that implements the SDL_gpu calls in terms of SDL_render. It's not very efficient but it works and it means we don't have to juggle two implementations of the main code.

BrettDong commented:
Sorry I build CDDA with homebrewed SDL in /usr/local/include &
/usr/local/lib, and I totaly have no experience with framework SDL and
no idea on how to update Makefile to build with framework SDL2_gpu. Do
you know how to build?


在 2016/11/25 上午12:43, patternoia 写道:
>
> Great work @mutability <https://github.com/mutability> !
> I need to warn you that if merged, this will break the macOS builds,
> because they are built with FRAMEWORK=1.
> |sdl-gpu| already supports building the |SDL2_gpu.framework| if build
> with |cmake -G "Unix Makefiles"|.
> You would need to update the |Makefile| accordingly to link against
> this framework like it is now done for regular |SDL2.framework|
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/CleverRaven/Cataclysm-DDA/pull/19412#issuecomment-262815177>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AUGWLuq3DMeyRPLdTbpu9wmcO8qugDzQks5rBb6sgaJpZM4K6BoA>.
>

patternoia commented:
I'll then provide the patch for frameworks next week if no one will do it faster.

mutability commented:
Seeing something odd with the alpha channel on semitransparent sprites - most obvious with gas fields (bloated zombie) - my current guess is that we're doing everything right but the driver is picking a 1-bit alpha channel format or something silly, so it's either fully transparent or fully opaque.

Trying some SDL_gpu changes to tell it to use 8-bit alpha.

jasonmbrown commented:
I'm really looking forwards to this merge, any updates on when that might be?

mutability commented:
Mostly waiting on build updates for macos & windows, neither of which I can do myself.
The CI environment will need a copy of SDL_gpu too, but it is in flux at the moment.
& if dropping software rendering support is no good, then it'll be a week+ while I sort that out

patternoia commented on src/cata_tiles.h:
```diff
@@ -3,6 +3,7 @@
 
 #include <SDL.h>
 #include <SDL_ttf.h>
+#include <SDL/SDL_gpu.h>
```
Could this be change just to `#include <SDL/SDL_gpu.h>` like it is done for `#include <SDL.h>`?
This breaks building with frameworks.

mutability replied:
This is a reasonable change, but it will need some work on other platforms to work correctly (ideally, pkg-config support upstream, I'll see what I can do)

patternoia commented:
When built with frameworks app hangs indefinitely on start with 100% cpu usage.
An endless loop?
```
$ lldb Cataclysm.app/Contents/Resources/cataclysm-tiles 
(lldb) target create "Cataclysm.app/Contents/Resources/cataclysm-tiles"
Current executable set to 'Cataclysm.app/Contents/Resources/cataclysm-tiles' (x86_64).
(lldb) env DYLD_LIBRARY_PATH=./Cataclysm.app/Contents/Resources/
(lldb) env DYLD_FRAMEWORK_PATH=./Cataclysm.app/Contents/Resources/
(lldb) r
Process 7917 launched: '/Users/pat/Cataclysm-DDA_source/Cataclysm.app/Contents/Resources/cataclysm-tiles' (x86_64)
Process 7917 stopped
* thread #1: tid = 0x62ce5, 0x000000010314eb23 SDL2`___lldb_unnamed_symbol1209$$SDL2 + 355, stop reason = signal SIGSTOP
    frame #0: 0x000000010314eb23 SDL2`___lldb_unnamed_symbol1209$$SDL2 + 355
SDL2`___lldb_unnamed_symbol1209$$SDL2:
->  0x10314eb23 <+355>: movl   (%r12), %edi
    0x10314eb27 <+359>: movl   0x18(%rdx), %eax
    0x10314eb2a <+362>: andl   %edi, %eax
    0x10314eb2c <+364>: movzbl 0x27(%rdx), %ebx
(lldb) bt
* thread #1: tid = 0x62ce5, 0x000000010314eb23 SDL2`___lldb_unnamed_symbol1209$$SDL2 + 355, stop reason = signal SIGSTOP
  * frame #0: 0x000000010314eb23 SDL2`___lldb_unnamed_symbol1209$$SDL2 + 355
    frame #1: 0x00000001031301e8 SDL2`___lldb_unnamed_symbol1033$$SDL2 + 328
    frame #2: 0x000000010315d0a7 SDL2`___lldb_unnamed_symbol1304$$SDL2 + 423
    frame #3: 0x00000001032a469a SDL2_gpu`copySurfaceIfNeeded + 106
    frame #4: 0x000000010328f478 SDL2_gpu`UpdateImage + 104
    frame #5: 0x00000001032326f8 SDL2_gpu`GPU_UpdateImage + 104
    frame #6: 0x000000010022a97c cataclysm-tiles`Atlas::add_sprite(this=0x000000010409ddf8, surface=0x000000010366cee0, source_rect=0x00007fff5fbf8470) + 1884 at cata_tiles.cpp:303
    frame #7: 0x0000000100232ce0 cataclysm-tiles`cata_tiles::load_tileset(this=0x000000010409ddf0, img_path="gfx//MShock32TilesetModded/tiles.png", R=-1, G=-1, B=-1, sprite_width=32, sprite_height=32) + 5184 at cata_tiles.cpp:683
    frame #8: 0x0000000100235961 cataclysm-tiles`cata_tiles::load_tilejson_from_file(this=0x000000010409ddf0, tileset_dir="gfx//MShock32TilesetModded", f=0x00007fff5fbfa788, image_path="gfx//MShock32TilesetModded/tiles.png") + 6337 at cata_tiles.cpp:770
    frame #9: 0x00000001002313b6 cataclysm-tiles`cata_tiles::load_tilejson(this=0x000000010409ddf0, tileset_root="gfx//MShock32TilesetModded", json_conf="tile_config.json", image_path="tiles.png") + 4534 at cata_tiles.cpp:716
    frame #10: 0x000000010022f10a cataclysm-tiles`cata_tiles::init(this=0x000000010409ddf0) + 4442 at cata_tiles.cpp:487
    frame #11: 0x0000000101ab4e6c cataclysm-tiles`curses_init() + 24396 at sdltiles.cpp:1612
    frame #12: 0x000000010053857d cataclysm-tiles`initscr() + 13 at cursesport.cpp:49
    frame #13: 0x0000000100d934b6 cataclysm-tiles`main(argc=0, argv=0x00007fff5fbff7a0) + 6550 at main.cpp:417
    frame #14: 0x00000001000014d4 cataclysm-tiles`start + 52
```

patternoia commented:
Given the issue above will be fixed, I confirm I am able to build cataclysm with sdl_gpu
Here is the OSX makefile patch

```
diff --git a/Makefile b/Makefile
index 7df1f50..805341e 100644
--- a/Makefile
+++ b/Makefile
@@ -450,9 +450,12 @@ ifdef TILES
       OSX_INC = -F$(FRAMEWORKSDIR) \
 		-I$(FRAMEWORKSDIR)/SDL2.framework/Headers \
 		-I$(FRAMEWORKSDIR)/SDL2_image.framework/Headers \
-		-I$(FRAMEWORKSDIR)/SDL2_ttf.framework/Headers
+		-I$(FRAMEWORKSDIR)/SDL2_ttf.framework/Headers \
+		-I$(FRAMEWORKSDIR)/SDL2_gpu.framework/Headers
       LDFLAGS += -F$(FRAMEWORKSDIR) \
-		 -framework SDL2 -framework SDL2_image -framework SDL2_ttf -framework Cocoa
+		 -framework SDL2 -framework SDL2_image -framework SDL2_ttf -framework Cocoa \
+		 -framework SDL2_gpu -framework OpenGL
+
       CXXFLAGS += $(OSX_INC)
     else # libsdl build
       DEFINES += -DOSX_SDL2_LIBS
@@ -826,6 +829,7 @@ ifdef FRAMEWORK
 	cp -R $(FRAMEWORKSDIR)/SDL2.framework $(APPRESOURCESDIR)/
 	cp -R $(FRAMEWORKSDIR)/SDL2_image.framework $(APPRESOURCESDIR)/
 	cp -R $(FRAMEWORKSDIR)/SDL2_ttf.framework $(APPRESOURCESDIR)/
+	cp -R $(FRAMEWORKSDIR)/SDL2_gpu.framework $(APPRESOURCESDIR)/
 ifdef SOUND
 	cp -R $(FRAMEWORKSDIR)/SDL2_mixer.framework $(APPRESOURCESDIR)/
 	cd $(APPRESOURCESDIR)/ && ln -s SDL2_mixer.framework/Frameworks/Vorbis.framework Vorbis.framework
```

mutability commented:
> When built with frameworks app hangs indefinitely on start with 100% cpu
usage.
> An endless loop?

Don't know what this is. Can you debug? Is it hung entirely within
updateimage or is the loop at a higher level?

mutability commented:
Also it would be useful to have symbols for SDL2

mutability commented:
Thirdly can you leave it running for, like, an hour and see if it resolves?

SDL_gpu appears to have a bug where if you tell it to update a small region of an image from a small region of a SDL_Surface, and the surface format is incompatible with the target image format, then it copies and converts the _entire_ source surface, not just the source region. In the CDDA case it is doing this for lots of 32x32 regions out of a huge image (the original tileset atlas image, which is typically something like 512x6000) and so it could just be copying/converting a huge image a few thousand times and taking a long time about it.

patternoia commented:
Yep, it started after roughly 20 minutes

patternoia commented:
> Also it would be useful to have symbols for SDL2
Am I overseeing a debug distribution of SDL2 somewhere? I cannot compile one from the source due to X11 errors I can not resolve

mutability commented:
I don't know about SDL2 on macos, sorry. Don't worry about the symbols, if it did eventually start then I'm pretty sure I know what's going on there.

mutability commented:
@patternoia can you check again with the commits I just pushed?

patternoia commented:
It works now. And I got 137 indoors and 102 fps outside the evac shelter on my 2011 laptop wows

patternoia commented:
So I put the `SDL_gpu.framework` on the OSX Jenkins CI build machine. When this pr will be merged, everything will be in place.

mutability commented:
Rebased onto current head and cleaned up the makefile a little.
This is as ready as I can make it.

Coolthulhu commented:
Do we have anyone with Windows+SDL to test this?

Will this build and work without SDL_gpu installed? If not, it needs to be a compile switch.

mutability commented:
This requires SDL_gpu. It is an invasive set of changes so a compile switch is not really practical, the closest you could easily do is to preserve a copy of the old sdltiles/cata_tiles and compile one or the other. I don't think that is a good idea.

mutability commented:
If you mean "can someone run the binaries without installing SDL_gpu separately" then yes, they can, for linux builds (SDL_gpu is linked staticly) and macos (the build includes sdl_gpu as a framework). MXE windows build should also be linking staticly IIRC. I don't know about other windows builds.

DanmakuDan commented:
> Do we have anyone with Windows+SDL to test this?

I was able to build it by adding libSDL2_gpu.dll.a to my linker options in CodeBlocks (most of the libraries in the project file are dynamically linked). SDL_gpu was built using MSYS2/mingw64.

kevingranade commented:
I really appreciate all the work on this, but I suspect this:
> The tiles version will only work if you have OpenGL 2 or higher available

May be a deal-breaker.

I'm particularly concerned  about users on older operating systems and/or
hardware no longer being able to use tiles.

We will need either support for software rendering or extensive feedback
from users that this new build works on their systems.

mutability commented:
I have a prototype software renderer under SDL_gpu but I would like to see this PR merged before I put more work in there.

kevingranade commented:
We are not going to break the game for an unknown number of users for an indeterminate period of time in order to try and merge this feature.
This is not merging without some strong indications that it will continue working on all the systems it currently works on.

mutability commented:
If that's the policy, ok, but I would point out that it certainly hasn't been applied elsewhere and requiring everything to be in huge monolithic PRs puts a much higher barrier on actually getting development done.

mugling commented:
Do we even need a software renderer - OpenGL 2.0 is a 12+ year old standard that a cursory search shows is supported by almost all hardware released in the last decade?

I can see updating the build process to get this compiling on all of the platforms we target being more an issue than actual hardware support.

kevingranade commented:
On Dec 7, 2016 9:43 AM, "mugling" <notifications@github.com> wrote:

Do we even need a software renderer - OpenGL 2.0 is a 12+ year old standard
that a cursory search shows is supported by almost all hardware released in
the last decade?

I can see updating the build process to get this compiling on all of the
platforms we target being more an issue than actual hardware support.

I could see that being the case as well, but I'm not certain enough of it
to remove software rendering support from the codebase.

All I'm asking for here is some due dilligence, get some builds out there
and demonstrate that people can run this new version, if you put a build
out and I get a big pile of, "works for me" and no reports of, "this
doesn't work", we're good to go.

Alternately, if you want to do this verification via the experimental
releases, there needs to be a working fallback of some kind so people can
keep playing in case they don't meet the new requirements.

macrosblackd commented:
There are players who have issues running tiles without the software
renderer.

On Wed, Dec 7, 2016 at 5:17 PM, Kevin Granade <notifications@github.com>
wrote:

> On Dec 7, 2016 9:43 AM, "mugling" <notifications@github.com> wrote:
>
> Do we even need a software renderer - OpenGL 2.0 is a 12+ year old standard
> that a cursory search shows is supported by almost all hardware released in
> the last decade?
>
> I can see updating the build process to get this compiling on all of the
> platforms we target being more an issue than actual hardware support.
>
> I could see that being the case as well, but I'm not certain enough of it
> to remove software rendering support from the codebase.
>
> All I'm asking for here is some due dilligence, get some builds out there
> and demonstrate that people can run this new version, if you put a build
> out and I get a big pile of, "works for me" and no reports of, "this
> doesn't work", we're good to go.
>
> Alternately, if you want to do this verification via the experimental
> releases, there needs to be a working fallback of some kind so people can
> keep playing in case they don't meet the new requirements.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/CleverRaven/Cataclysm-DDA/pull/19412#issuecomment-265604630>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABUR_im80bRJD9kxjBGQ9XwAuGpKSMWyks5rFz5wgaJpZM4K6BoA>
> .
>

DanmakuDan commented:
Related to #13863, software rendering is used because the graphics card didn't support render targets. For Windows machines, SDL will default to using DirectX as the renderer. The graphics card either falsely reports that render targets are valid, or that render targets don't specifically work with DirectX.

> Updated windows builds with the wide-character fix and a faster minimap:
> 
> https://www.dropbox.com/s/ici4v4ih3dan0k9/cdda-sdlgpu-win-5dc16ff6.zip?dl=0 (full package)
> https://www.dropbox.com/s/p4a0ut4e6f64gq7/cataclysm-tiles-5dc16ff6.exe?dl=0 (just the exe)

The builds uploaded here could be linked on the forums, so that possibly someone with an older computer can check if the OpenGL renderer has better luck with the render target issue.

mutability commented:
Here is my WIP on the software renderer (a bit glitchy, tile border alignment and clipping are off)

https://github.com/mutability/Cataclysm-DDA/tree/sdlgpu-software-rendering

kevingranade commented:
I'd LOVE to have faster tiles rendering, but I have significant concerns (in order of greatest to least severity).
It kills the software rendering fallback, the concern here is that poor hardware support or misconfiguration will be escalated to "you can't run the game with tiles".
It adds a new mandatory library dependency for tiles builds, the problem here is it imposes additional overhead on developers who want to get started with making code changes, not being able to get a build working drives off developers.
The dependency isn't on a standard library, but a one-off wrapper library being pulled from a github repository.  This isn't a dealbreaker, but it does exacrbate the pevious issue.
This change is huge and hard as hell to review.

Condidering how invasive this is, it really makes me wonder where that speedup we're seeing is coming from and if taking this dependency is the easiest way to get it.  The readme about the library gives top billing to it's draw batching feature, and from what I can see that is literally the only feature of the library we will be using that is distinct from SDL itself.  This makes me wonder how hard it would be to integrate draw operation batching into our application, either as a wrapper or integrated into the draw code.  This would also have the benefit of speeding up software rendering.  It seems at least possible to me that this would be a 1KLoC or so feature, weighing in at significantly less code than what is required to integrate with this library.

Setting that possibility aside, the above issues need to be mitigated somehow.
I'm not comfortable pushing this code with no software rendering fallback, we get a significant number of reports about rendering isues just based on when the automatic fallback fails, there is no telling how often users have the fallback work, in which case this version of the code wold simply fail for them.  As I mentioned earlier, a large enough testing campaign could have the same effect, but I haven't seen anything happening in that drection.
I really do not like the hard dependency on a new library, especially one that you have to build locally.  This isn't an absolute, but a buid-time fallback to jut SDL would be a very good thing to have, even if it's a messy pile of #ifdefs.
Ideally this would be broken into several PRs each of which has an outline of its rationale and is seperately reviewable and buildable, even if they're untestable, e.g. adding the atlas class without using it.  If that's not possible or doesn't get us to a small enough core change, it needs a lot more explanation of what all the code is doing.  In it's current state reviewing the code seems to require replicating all of the work that went into you implementing it in the first place, which is fine with smaller changes but a non-starter for changes this extensive.

mutability commented:
Unfortunately, I am unlikely to do any of that.

If you want to eliminate the development-time requirement to build SDL_gpu (there is no runtime requirement) then you could include a copy of it in the source tree as is done for some other code.

If you want to remove the use of SDL_gpu and write your own tile batcher, that'd be great, but that's not something I will be doing. I don't have the necessary opengl knowledge and I don't really want to acquire it right now; and opengl is notoriously picky to get right across lots of systems. The reason I picked SDL_gpu is exactly so that we didn't need to rediscover that knowledge.

If you want to organize some more widespread testing, that'd be great, but it's not something I can really do.

I am very, very unlikely to split this PR into multiple PRs given the past history of trying to get anything merged into this project. I would guess that it would take 6 months or more of wrangling PRs to get that done. I already have other changes that require this PR to be committed that are on hold.

I took care to split the changes into logically separate _commits_ that should individually work - you should review those.

I can't do much about it if you just don't understand the changes. If some areas need more comments, let me know where.

mutability commented:
The software rendering branch linked above basically works except for some tile alignment glitches that are looking like a SDL_render bug. I would want everything else sorted out before spending more time on that.

mutability commented:
Rebased to current master, waiting on a merge or feedback to my comment above.

Updated, untested windows build: https://www.dropbox.com/s/fovgy4bv347g2c5/cdda-sdlgpu-win-922a5cfa.zip?dl=0

kevingranade commented:
There's no feedback to give, "these things must happen for this to be merged", "I won't do those things".  The obvious result is no merge.
Anyone can feel free to reopen if they want to meet the above requirements.

mutability commented:
Please do me the courtesy of actually reading and responding to my comment above; I made several suggestions about how to make progress on this issue.

kevingranade commented:
I had no intention of being rude, I took this:
> Unfortunately, I am unlikely to do any of that.

To mean your post was suggestions for other people to take up, not
yourself, and that no detailed response was needed.  Sorry for
misunderstanding.

Importing the SDL_gpu source partially addresses the dependency issue, but
is a poor practice and also adds a hard build dependency on cmake, which is
unacceptable.  On the off chance you mean including a SDL_GPU binary,
that's simply not possible due to our variety of build targets and shifting
compiler options.

A native tile batching solution would be independent of this pr so is
irrelevant to it staying open.

I have no interest in running a campaign to determine whether this change
does what I expect it to by failing to load in the face of poor hardware
support.  If I were to start working on this change, it would be to
fragment it into manegable chunks that don't disable the existing
functionality.  I have no intention of doing this in the near future.

As for pr organization, I want you to break it up and comment it more
before I review it, you want me to review it and tell you what needs
improvement, it's an impasse.

mutability commented:
Well, I tried. I just don't see how multiple PRs would work, TBH: each PR would include all changes of the dependent PRs and it would become a rebase nightmare. I have some time to spare to work on this, but not that much time! 

Like you, I have no interest in running a testing campaign.

I'll just rebase this against master periodically and put up new builds and hope it gets some organic testing.

mutability commented:
I can do a naive decomposition into half a dozen PRs if you think that's really a useful thing - one commit per PR.

kevingranade commented:
I'm not interested in discussing this unless there's a solution available for both of the major issues.

1. Software rendering must be supported.
2. SDL_gpu must be an optional build dependency.

mutability commented:
(1) I can do - see the branch I linked above.
(2) you said wasn't a dealbreaker, has something changed since your previous comment?

mutability commented:
For (2) would you accpet "include a copy of SDL_gpu"? (The cmake dependency is easy to break).

I would prefer "include a copy" over "two implementations", as if we have two parallel implementations only one of which is actually built in the CI builds, then (a) the pure SDL implementation is going to rot and (b) we cannot do things that SDL_gpu makes possible but which are not easy to do in vanilla SDL (e.g. premultiplied alpha)

mutability commented:
Rebased, updated windows builds (untested): https://www.dropbox.com/s/scco5axx4blwkfn/cdda-sdlgpu-win-31d99bde.zip?dl=0

@kevingranade: Would you accept "include a copy of SDL_gpu"?

kevingranade commented:
> I'm not interested in discussing this unless there's a solution available for both of the major issues.

>    Software rendering must be supported.
>    SDL_gpu must be an optional build dependency.

> Importing the SDL_gpu source partially addresses the dependency issue, but is a poor practice and also adds a hard build dependency on cmake, which is unacceptable.

mutability commented:
So is importing sdlgpu and removing the cmake dependency something that you
would accept or not? I still don't know! I asked the question because your
comment was ambiguous, just quoting it back at me does not help me make any
progress here.

I am not being obtuse here, I just do not want to start down that path only
to find out later was not something you would accept.

You are making it very difficult to contribute code to the project.

kevingranade commented:
On Jan 23, 2017 3:43 AM, "Oliver Jowett" <notifications@github.com> wrote:

So is importing sdlgpu and removing the cmake dependency something that you
would accept or not?


As stated, no. Importing code from another project is pretty iffy already,
you're suggesting making a fork of the project, which means we would have
to keep updating this new build system if we want to keep in sync with
upstream.
If you convince upstream to drop the cmake dependency, that'd be fine, but
it's also incredibly unlikely.

You are making it very difficult to contribute code to the project.


I'm making it difficult to take a new, HARD dependency on a graphics
library I've never heard of before, in a project that has been
intentionally and explicitly dependency averse for its entire history.

mutability commented:
I mean you are making it organizationally hard. If I go ahead and do my own design, it gets vetoed after the fact. If I ask for input, it takes a month and quite a bit of fighting to get even a minimal response.

I don't think we are going to reach agreement on this particular issue, since the changes do fundamentally require sdlgpu, but you don't want the dependency. I don't see maintaining two copies of the graphics code as practical. So this PR is effectively dead.

I will maintain the changes in my own repo.
